src/core/task/task.h

```diff
@@ -96,9 +98,26 @@ class LegateTask {
     Core::show_progress(task, legion_context, runtime, task_name());
 
     TaskContext context(task, regions, legion_context, runtime);
-    if (!Core::use_empty_task) (*TASK_PTR)(context);
 
-    return context.pack_return_values();
+    try {
+      if (!Core::use_empty_task) (*TASK_PTR)(context);
+      return context.pack_return_values();
+    }
+    // abstract class cannot be re-thrown, so we have to enumerate all concrete Realm exceptions
+    catch (Realm::CancellationException& e) {
+      throw e;
+    } catch (Realm::PoisonedEventException& e) {
+      throw e;
+    } catch (Realm::ApplicationException& e) {
+      throw e;
+    } catch (legate::TaskException& e) {
+      if (context.can_raise_exception()) {
+        context.make_all_unbound_stores_empty();
+        return context.pack_return_values_with_exception(e.index(), e.error_message());
+      } else
+        // This is an unexpected exception so better be rethrown
+        throw e;
+    }
```

Comment:
You may be able to use just `throw;` which reuses the same exception object, and can be used with abstract classes:

```
      catch (const legate::TaskException& e) {
        if (context.can_raise_exception()) {
          context.make_all_unbound_stores_empty();
          return context.pack_return_values_with_exception(e.index(), e.error_message());
        } else
          // This is an unexpected exception so better be rethrown
          throw;
      } catch (const std::exception& e) {
        throw;
      }
```

Come to think of it, do you need to handle all other exception classes here explicitly? Assuming the exception doesn't match `legate::TaskException`, won't it automatically be re-thrown to the parent function?

Reply:
Initially I was trying to catch all non-Legate C++ exceptions except the Realm ones and re-throw them on the Python side with a generic Legate exception. Then I realized that that would effectively require the core to do exception handling for every task, which I don't think we want. So I gave up on that idea but forgot to remove the cases for Realm exceptions that are obsolete.

I updated the code so that it only handles Legate exceptions. In case they are raised by tasks that did not declare any Python exceptions, they are reported as errors now.